### PR TITLE
Use OAuth for authentication.

### LINKED ISSUE #5: Update to Guzzle 6
The latest, the greatest, the guzzliest.

### LINKED ISSUE #13: Support OAuth 2 Password Grant.
🔏

### LINKED ISSUE #14: Support OAuth 2 client credentials grant.
🔐

### PR DESCRIPTION
**Pull request by DFurnes**
#### Changes

Closes #13. Closes #14. This PR adds support for Northstar's OAuth 2 [password grant](https://github.com/DoSomething/northstar/blob/dev/documentation/endpoints/oauth.md#create-token-password-grant) and [client credentials grant](https://github.com/DoSomething/northstar/blob/dev/documentation/endpoints/oauth.md#create-token-client-credentials-grant), by way of a new `AuthorizesWithNorthstar` trait for the `RestApiClient`. In the future, this will allow any other PHP API clients we make using `RestApiClient` to also easily sign requests with tokens from Northstar.

Also closes #5. 🚰
#### What's ~~not~~ done?
- [x] Handle cases where token is refused unexpectedly.
- [x] Handle cases when provided refresh token is invalid.
- [x] Revoke refresh token on logout.

---

For review: @angaither @weerd

### CODE REVIEW

**Comment by DFurnes** [comment]
Oy, this one turned into a monster. Should be ready for review now. 😬

**Comment by sheyd** [comment]
cb-froghemoth

File: src/Common/RestApiClient.php [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+     */
+    protected function getAuthorizationHeader()
+    {
+        return null;
+    }
+
+    /**
+     * Handle unauthorized exceptions.
+     *
+     * @param $endpoint - The path that
```
Comment (weerd) [inline src/Common/RestApiClient.php]: to be continued? (incomplete sentence :)

File: src/Common/RestApiClient.php [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+     * @throws ForbiddenException
+     * @throws InternalException
+     * @throws UnauthorizedException
+     * @throws ValidationException
+     */
+    public function send($method, $path, $options = [], $withAuthorization = true)
+    {
+        try {
+            // Increment the number of attempts so we can eventually give up.
+            $this->attempts++;
```
Comment (weerd) [inline src/Common/RestApiClient.php:171]: Nice :)

File: src/Common/RestApiClient.php (hunk 1 above)
Comment (DFurnes) [inline src/Common/RestApiClient.php]: The path _that_. 😤

(Yeah.)

Comment (DFurnes) [inline src/Common/RestApiClient.php]: Fixed in e458529!

File: src/NorthstarClient.php [hunk 3]
```diff
@@ -54,7 +39,7 @@ public function verify($credentials)
      */
     public function getAllUsers($inputs = [])
     {
-        $response = $this->get('users', $inputs);
+        $response = $this->get('v1/users', $inputs);
```
Comment (weerd) [inline src/NorthstarClient.php:42]: Not sure, but would it make sense at all to store the version number as a property or config setting somewhere?

Comment (weerd) [inline src/NorthstarClient.php:42]: Oh maybe not, since you later specify two different versions.

**Comment by weerd** [comment]
Gah! So. Much. Code. Haha. Looks good! Nothing that stuck out that seemed super questionable. Nice work Mr. FUUUUUURNS! 

:sheep: it!

File: src/NorthstarClient.php (hunk 3 above)
Comment (DFurnes) [inline src/NorthstarClient.php:42]: Yep, we've got a mix of v1 and v2 endpoints right now. 😁

**Comment by DFurnes** [comment]
🙇

**Pull request merged**